src/Features/CSharp/Portable/ConvertForEachToFor/CSharpConvertForEachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,178 @@
+            var collectionVariable = GetCollectionVariableName(
+                model, generator, foreachInfo, foreachCollectionExpression, cancellationToken);
+
+            var collectionStatementType = foreachInfo.RequireExplicitCastInterface ?
+                GetTypeExpression(generator, foreachInfo.Options, foreachInfo.ExplicitCastInterface) :
+                GetTypeExpression(generator, foreachInfo.Options,
+                    model.GetTypeInfo(foreachCollectionExpression).Type ?? model.Compilation.GetSpecialType(SpecialType.System_Object));
+
+            // first, see whether we need to introduce new statement to capture collection
+            IntroduceCollectionStatement(
+                model, foreachInfo, editor, collectionStatementType, foreachCollectionExpression, collectionVariable);
+
+            var indexVariable = CreateUniqueName(foreachInfo.SemanticFacts, model, foreachStatement.Statement, "i", cancellationToken);
+
+            // put variable statement in body
+            var bodyStatement = GetForLoopBody(generator, foreachInfo, collectionVariable, indexVariable);
+
+            // create for statement from foreach statement
+            var forStatement = SyntaxFactory.ForStatement(
+                SyntaxFactory.VariableDeclaration(
+                    GetTypeExpression(
+                        generator, foreachInfo.Options, model.Compilation.GetSpecialType(SpecialType.System_Int32)),
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                            indexVariable.WithAdditionalAnnotations(RenameAnnotation.Create()),
+                            argumentList: default,
+                            SyntaxFactory.EqualsValueClause((ExpressionSyntax)generator.LiteralExpression(0))))),
+                SyntaxFactory.SeparatedList<ExpressionSyntax>(),
+                (ExpressionSyntax)generator.LessThanExpression(
+                    generator.IdentifierName(indexVariable),
+                    generator.MemberAccessExpression(collectionVariable, foreachInfo.CountName)),
+                SyntaxFactory.SingletonSeparatedList<ExpressionSyntax>(
+                    SyntaxFactory.PostfixUnaryExpression(
+                        SyntaxKind.PostIncrementExpression, SyntaxFactory.IdentifierName(indexVariable))),
+                bodyStatement);
+
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                // move comments before "foreach" keyword to "for". if collection statement is introduced,
+                // it should have attached to the new collection statement, so no need to do it here.
+                forStatement = forStatement.WithLeadingTrivia(foreachStatement.GetLeadingTrivia());
+            }
+
+            // replace close parenthese from "foreach" statement
+            forStatement = forStatement.WithCloseParenToken(foreachStatement.CloseParenToken);
```

Comment:
1. this seems strange.  why do this here.  why not just pass along *all* parens when creating the forStatement?
2. what about hte open paren?

Reply:
comment on those are explicitly cut in this PR. will have other PR on how to handle comments. already talked with people, and okay on this iteration.

Reply:
I guess  i don't understand why you'd do that.  You have the parens from the previous statement.  THe new statement needs parens.  It's *simpler* to just pass the parens from one to the other (and, as a bonus, everything works better than this code).  So we've gone a more complex route to get a less successful outcome.  That doesn't make sense to me.  We could just fix this code to pass along the parens simply.

Note: this fits the conversation about principles where we shoudl def try to preserve comments that are trivial to preserve when the before/after can trivially share the same components.

Reply:
This still applies